refactor(lidar_centerpoint): rework parameters

tby-udel opened the pull request:
## Description
Implement the ROS Node configuration layout described in https://github.com/orgs/autowarefoundation/discussions/3371 for the `lidar_centerpoint` package.

 * Move the default parameter file to a standardized location. Use the JSON Schema for that.
 * Updated readme file.

## Tests performed
Package built and launched locally.

## Notes for reviewers
 Nothing inside the src code part was changed, only taking the parameters out.

## Interface changes
Parameter values are now required to be passed to the node when launched.

## Effects on system behavior
None

## Pre-review checklist for the PR author
The PR author **must** check the checkboxes below when creating the PR.

 * [x]  I've confirmed the [contribution guidelines](https://autowarefoundation.github.io/autoware-documentation/main/contributing/).
 * [x]  The PR follows the [pull request guidelines](https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/).

 ## In-review checklist for the PR reviewers
 The PR reviewers **must** check the checkboxes below before approval.

 * [ ]  The PR follows the [pull request guidelines](https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/).
 * [ ]  The PR has been properly tested.
 * [ ]  The PR has been reviewed by the code owners.
 
 ## Post-review checklist for the PR author
 The PR author **must** check the checkboxes below before merging.
 
 * [ ]  There are no open discussions or they are tracked via tickets.
 * [ ]  The PR is ready for merge.
 
 After all checkboxes are checked, anyone who has write access can merge the PR.

tby-udel commented:
![image](https://github.com/autowarefoundation/autoware.universe/assets/144167190/90b1653e-6ba3-4d98-b9a4-0fdb581fe57f)
In line 105, the check passed, as centerpoint_tiny_ml_package.param.yaml should be compared with centerpoint_tiny_ml_package.schema.json. However, in line 106, it's centerpoint_tiny_ml_package.param.yaml comparing with centerpoint_tiny.schema.json, which should be happening, they are meant for different files. 
The line 115 error has been corrected by deleting the 'has_twist' attribute from .schema.json file.

knzo25 commented:
@tby-udel 
It seems that behavior is by design
https://github.com/autowarefoundation/autoware-github-actions/blob/main/json-schema-check/validate_json_schemas.py

What happens is that first `perception/lidar_centerpoint/schema/centerpoint_tiny.schema.json` is selected as the target for the check. Then, all `.param.yaml` files starting with `centerpoint_tiny` are checked, which end up with two parameter files being tested, among them one that should not be checked. We probably can not change the github action, so one approach would be not use `centerpoint_tiny` but `centerpoint_tiny_base` or another name that eliminates the base name conflict

tby-udel commented:
I have renamed the files and made necessary changes to the yaml,schema and launch files.

knzo25 commented:
@tby-udel 
I think that I did not correctly conveyed my intention in the previous post after seeing the latest changes.

The reason it fails is that
`centerpoint.schema.json` triggers checks for `centerpoint.param.yaml` and `centerpoint_ml_package.param.yaml` since they share the same `centerpoint` name.
What I proposed is to have `centerpoint_base` and `centerpoint_ml_package` so that there are no more conflicts and each json would only trigger one yaml check.

The change to "_base" should not propagate in more places that the name of the param file

tby-udel commented:
If the parameter file's name is changed, then the launch file won't be able to find the correct parameter file, as the parameter file is used through:

  arg name="data_path" default="$(env HOME)/autoware_data" description="packages data and artifacts directory path"
  **arg name="model_name" default="centerpoint_tiny" description="options: `centerpoint` or `centerpoint_tiny`"**
  arg name="model_path" default="$(var data_path)/lidar_centerpoint"
  **arg name="model_param_path" default="$(find-pkg-share lidar_centerpoint)/config/$(var model_name).param.yaml"**
  arg name="ml_package_param_path" default="$(var model_path)/$(var model_name)_ml_package.param.yaml"
  arg name="class_remapper_param_path" default="$(var model_path)/detection_class_remapper.param.yaml"
  arg name="build_only" default="false" description="shutdown node after TensorRT engine file is built"

In order to correctly launch this node, if the parameter file's name is changed to centerpoint_base.param.yaml, then inside the launch file, for the "**model_name**" part,it should be **description="options: `centerpoint_base` or `centerpoint_tiny`** , and only after changing the launch file, can the "**model_param_path**" become the correct **/config/$(var model_name).param.yaml**.

And in the current centerpoint.param.yaml file, the onnx paths and engine paths are organized as this:

encoder_onnx_path: "$(var model_path)/pts_voxel_encoder_**$(var model_name)**.onnx"
encoder_engine_path: "$(var model_path)/pts_voxel_encoder_**$(var model_name)**.engine"
head_onnx_path: "$(var model_path)/pts_backbone_neck_head_**$(var model_name)**.onnx"
head_engine_path: "$(var model_path)/pts_backbone_neck_head_**$(var model_name)**.engine"

So now that the model_name is changed from centerpoint to centerpoint_base, but these files' name are still centerpoint.onnx and centerpoint.engine, so instead of **$(var model_name)**, they should all be changed to **centerpoint**.

If the json_schema_check cannot change, I think we can't solve this problem by changing the centerpoint.param.yaml only, so I changed centerpoint.param.yaml (renamed and changed the onnx/engine paths), centerpoint.schema.json(renamed) and the launch file.

knzo25 commented:
@tby-udel 
I think we still have a misunderstanding
`model_name` should still be `centerpoint` or `centerpoint_tiny`.

What should change is 
`<arg name="model_param_path" default="$(find-pkg-share lidar_centerpoint)/config/$(var model_name).param.yaml/>"`
to
`<arg name="model_param_path" default="$(find-pkg-share lidar_centerpoint)/config/$(var model_name)_base.param.yaml/>"

and so on.

If you think this is going to deep and out of scope for the contribution / time you expected, let us know and we can continue this from here

knzo25 commented:
@tby-udel 
If you need me to run the CI/CD please drop me a mention. Otherwise, there is a chance I will not see that you made a commit and want to try it (I just ran it, and it is not passing the tests :cry: )

tby-udel commented:
I have corrected a mistake in centerpoint_ml_package.json.schema file, and have pushed the signoff commit. Please try it again.

knzo25 commented:
@tby-udel 
It seems that the tests are passing now :tada: 

Some concerns though:
 - The changes to the file `perception/lidar_centerpoint/config/centerpoint_base.param.yaml` seem redundant since the ones for tiny were not needed. If that is the case, can you revert those changes (keep the contents of the original file)
 - In the readme, you added all the macros together, which will make it difficult to read. Can you add them in separate sections?
 - In the related `autoware_launch` project, the config files appear as well: https://github.com/autowarefoundation/autoware_launch/blob/main/autoware_launch/launch/components/tier4_perception_component.launch.xml
 - With regards to this last point, a corresponding PR should be sent there as well, and we would merge the PRs together to minimize discrepancies. Would you send a PR there as well? If not, we can send it :+1:

knzo25 reviewed (changes requested):
@tby-udel 
It seems that the tests are passing now :tada: 

Some concerns though:
 - The changes to the file `perception/lidar_centerpoint/config/centerpoint_base.param.yaml` seem redundant since the ones for tiny were not needed. If that is the case, can you revert those changes (keep the contents of the original file)
 - In the readme, you added all the macros together, which will make it difficult to read. Can you add them in separate sections?
 - In the related `autoware_launch` project, the config files appear as well: https://github.com/autowarefoundation/autoware_launch/blob/main/autoware_launch/launch/components/tier4_perception_component.launch.xml
 - With regards to this last point, a corresponding PR should be sent there as well, and we would merge the PRs together to minimize discrepancies. Would you send a PR there as well? If not, we can send it :+1:

tby-udel commented:
I will start on your suggestions as soon as possible, I just came back from a lot of other stuff and July 4th National Day. Pardon my lateness.

tby-udel commented:
Hi Kenzo, it feels so good to find out that it finally successfully passed the json-schema-check. 

1. For the first point, do you mean that the 'centerpoint_tiny' file shouldn't  be attached with a '_base' ? If that is the case, I think it's necessary as the following screenshot suggests that the launch file is calling for two model names, either `centerpoint` or `centerpoint_tiny`, and we are adding '_base' to the base methods, so the 'centerpoint_tiny' will need the '_base'.
![image](https://github.com/autowarefoundation/autoware.universe/assets/144167190/4d95c4af-9a6f-44d7-9f6d-1d0dfd2f9f1a)

2. I have reorganized the readme file to make it more readable, and as the param.yaml files mainly have 2 different types of organization and 1 remapper, so they are now separated.

3. Could you help me send the corresponding PR to https://github.com/autowarefoundation/autoware_launch/blob/main/autoware_launch/launch/components/tier4_perception_component.launch.xml ? This is my first time using github, and I am still learning about how does the system work and I don't know where to put the PR link. Thank you so much for your help.

knzo25 commented:
@tby-udel 
With regards to the first point, I only meant that these changes do not seem to be needed
![image](https://github.com/autowarefoundation/autoware.universe/assets/4712270/97099c1c-9bdb-4580-851c-9b2bdcc40fce)

About the last point. Do you mean you would prefer to me to send the PR instead? if so, no problem ! (I would just like to have a confirmation)

tby-udel commented:
Hi Kenzo, requested changes are made. And yes I confirm that I would prefer you to help me send the PR instead. Thank you for your help.

knzo25 commented:
Sent the companion PR in:
https://github.com/autowarefoundation/autoware_launch/pull/1061

knzo25 commented:
@tby-udel 
I missed one file that needs to be addressed in this PR:
`launch/tier4_perception_launch/launch/object_recognition/detection/detector/lidar_dnn_detector.launch.xml`
is still using the old naming convention

tby-udel commented:
Requested changes for tier4_perception_launch and README.md have been made.

knzo25 commented on perception/lidar_centerpoint/README.md:
````diff
@@ -234,35 +222,16 @@ You can find it in `mmdetection3d/projects/AutowareCenterPoint` file.
-```
+#### Adjust the config file for the custom model
+
+All the ROS parameters have been moved into `.param.yaml` files, feel free to change the parameters inside them for better performance! **centerpoint\_(your_selection_of_model).param.yaml** files are under the config file directory of the lidar_centerpoint node. The information for these parameters are shown in the **Parameters** section.
````
**Parameters** section
should be instead a link in markdown style

knzo25 replied:
The change in this section is not adequate in my opinion. In the previous version we explained the process of creating a new model/version of centerpoint (i.e., how to create new parameters, modify launchers ,etc). This idea should be kept in this PR.

Expressions including exclamations are not adequate in this kind of documents

knzo25 commented on perception/lidar_centerpoint/README.md:
```diff
@@ -28,33 +28,21 @@ We trained the models using <https://github.com/open-mmlab/mmdetection3d>.
+
+### The `build_only` option
+
+The `lidar_centerpoint` node has `build_only` option to build the TensorRT engine file from the ONNX file, and the `build_only` option is inside the launch file of `lidar_centerpoint` node, and all the ROS parameters are moved in `.param.yaml` files.
```
```suggestion
The `lidar_centerpoint` node has a `build_only` option to build the required TensorRT engine file from an ONNX file. This parameter is set in the launch file of this package, whereas all other parameters are set via `.param.yaml` files.
```

knzo25 commented on launch/tier4_perception_launch/launch/object_recognition/detection/detector/lidar_dnn_detector.launch.xml:
```diff
@@ -3,7 +3,7 @@
   <arg name="lidar_detection_model" default="centerpoint" description="options: `transfusion`, `centerpoint`, `apollo`, `clustering`"/>
   <!-- Lidar detector centerpoint parameters -->
-  <arg name="centerpoint_model_name" default="centerpoint_tiny" description="options: `centerpoint`, `centerpoint_tiny` or `centerpoint_sigma`"/>
+  <arg name="centerpoint_model_name" default="centerpoint_tiny" description="options: `centerpoint`, `centerpoint_tiny` or `centerpoint_sigma` P.S. The _sigma model only has ml package version."/>
```
While we appreciate you pointing out to us the state of the sigma model, leaving P.S. in files is not a good practice. I will notify the maintainers of this model, but in the meantime remove this comment

knzo25 reviewed (changes requested):
Left some comments that need addressing

knzo25 commented:
@kminoda @YoshiRi 
It seems that there are some param files missing for the sigma centerpoint. How should we proceed with this?
Also please do not forget to review the companion / sister PR

tby-udel commented:
All the requested changes are made.

knzo25 commented on perception/lidar_centerpoint/README.md:
````diff
@@ -234,35 +222,16 @@ You can find it in `mmdetection3d/projects/AutowareCenterPoint` file.
-```
+#### Adjust the config file for the custom model
+
+All the ROS parameters have been moved into `.param.yaml` files, please set the parameters of the config file like point*cloud_range, point_feature_size, voxel_size, etc. in the config files. \*\*centerpoint*(your_selection_of_model).param.yaml\*\* files are under the config file directory of the lidar_centerpoint node. The information for these parameters are shown in the [Parameters](#parameter ...
````
Sorry, this lacks a little in both writing and content.
The idea of the section is to describe the process of creating the files for a new model. In addition to some grammatical errors, instead of giving examples of parameters of the config file explain what kind of parameters they have. This links to the the ml config file, which is not mentioned here

tby-udel replied:
> Sorry, this lacks a little in both writing and content. The idea of the section is to describe the process of creating the files for a new model. In addition to some grammatical errors, instead of giving examples of parameters of the config file explain what kind of parameters they have. This links to the the ml config file, which is not mentioned here

@knzo25 Thank you for your instructions, I have updated the README.md file, please check if this meets your expectation. If you can give proposal for me to accept, it would be of great help for me. I will click the button commit suggestion afterwards. I really learned a lot with your help for this PR, thank you so much.

knzo25 commented:
@tby-udel 
Since you mentioned you are new to git/github I think it is not uncalled for to mention that when a reviewer presents some `code suggestions`, and you agree with those, it is customary to accept the changes instead of making a new commit with those changes. It makes reviewing easier. In the same fashion, when you address a comment, you should paste the hash of the related commit in the conversation, to the reviewer can check that his/her concern was handled.

There are some parts of the documentation that are not very good and I left the explanation as a review. If you prefer that I give a proposal for you to accept it let me now ! (and in that case please click the button `commit suggestion`)

knzo25 commented:
@tby-udel 
There are some conflicts

knzo25 reviewed (changes requested):
@tby-udel 
There are some conflicts

tby-udel commented:
> @tby-udel There are some conflicts

I have merged the main branch into the current branch, there doesn't seem to be conflicts after merging.

tby-udel commented on perception/lidar_centerpoint/schema/centerpoint_base.schemal.json:
```diff
@@ -3,9 +3,29 @@
   "title": "Parameters for Lidar Centerpoint Node",
```
@kminoda I have fixed the name of the file.

Shin-kyoto commented:
@tby-udel 
Thank you for creating the PR.
Would it be possible for you to resolve the conflicts? 

To keep Autoware healthy, we close PRs if there is no response for more than a week after requesting conflict resolution. If this PR is still active, please let us know. Even if this PR is closed, you are welcome to reopen it once the conflicts have been resolved.

Shin-kyoto commented:
@tby-udel

friendly ping

Shin-kyoto commented:
@tby-udel

friendly ping

Shin-kyoto commented:
@tby-udel

friendly ping

Shin-kyoto commented:
@tby-udel

We did not get the reply from author for a month, so I close this PR.
@tby-udel, thank you so much for your PR and feel free to  reopen it.
